File: src/app/operations_dashboard.py
"""Small read-only dashboard contract. Never obtains broker data or starts work."""
from __future__ import annotations

from collections.abc import Mapping, Sequence
from datetime import datetime, timezone
from typing import Any


STAGE_LABELS = {
    "engine": "엔진", "live_armed": "주문 권한", "market_session": "거래 시간",
    "buy_candidates": "후보 종목", "micro_buy_intents": "진입 조건",
    "strategy_election": "전략 선택", "position": "포지션",
}


def _mapping(value: Any) -> dict[str, Any]:
    return dict(value) if isinstance(value, Mapping) else {}


def _order_rows(events: Sequence[Mapping[str, Any]]) -> list[dict[str, Any]]:
    """Normalize actual submission/status events; strategy signals are not orders."""
    accepted = {"SUBMITTED", "ACCEPTED", "OPEN", "WORKING", "FILLED", "PARTIALLY_FILLED", "REJECTED", "BLOCKED", "CANCELED", "CANCELLED", "ERROR"}
    normalized = []
    for event in events:
        if not isinstance(event, Mapping):
            continue
        symbol = event.get("ticker") or event.get("symbol")
        order_id = event.get("broker_order_id") or event.get("order_id")
        side = str(event.get("side") or event.get("kind") or "").upper()
        state = str(event.get("order_status") or event.get("status") or event.get("outcome") or "").upper()
        quantity = event.get("ordered_quantity", event.get("quantity"))
        if not symbol or state not in accepted:
            continue
        if not order_id and not (side in {"BUY", "SELL"} and quantity is not None):
            continue
        normalized.append({
            "occurred_at": event.get("occurred_at") or event.get("recorded_at") or event.get("at"),
            "ticker": symbol, "side": side if side in {"BUY", "SELL"} else None,
            "order_status": state, "broker_order_id": order_id,
            "ordered_quantity": quantity, "filled_quantity": event.get("filled_quantity"),
        })
    normalized.sort(key=lambda item: str(item.get("occurred_at") or ""))
    merged: dict[str, dict[str, Any]] = {}
    for index, row in enumerate(normalized):
        key = str(row.get("broker_order_id") or f"unsubmitted:{index}")
        merged[key] = {**merged.get(key, {}), **{k: v for k, v in row.items() if v is not None}}
    return sorted(merged.values(), key=lambda item: str(item.get("occurred_at") or ""))[-20:]


def build_operations_overview(
    *,
    trading: Mapping[str, Any] | None = None,
    blockade: Mapping[str, Any] | None = None,
    markets: Mapping[str, Any] | None = None,
    runtime: Mapping[str, Any] | None = None,
    mode: Any = None,
    orders: Sequence[Mapping[str, Any]] | None = None,
    now: datetime | None = None,
) -> dict[str, Any]:
    """Normalize cached snapshots; an unknown state never means permission.

    KR and US observations are independent. The existing single execution owner
    is explicitly reported rather than implying two simultaneous order engines.
    """
    trading = _mapping(trading)
    status = _mapping(trading.get("status")) or trading
    session = _mapping(status.get("strategy_session")) or _mapping(_mapping(status.get("last_summary")).get("strategy_session"))
    blockade = _mapping(blockade)
    chain = [dict(item) for item in blockade.get("chain", []) if isinstance(item, Mapping)]
    first_blocker = next((item for item in chain if item.get("ok") is not True), None)
    session_link = next((item for item in chain if item.get("stage") == "market_session"), {})
    scanned = _mapping(_mapping(session_link.get("data")).get("scanned_groups"))
    markets = _mapping(markets)
    market_rows = {}
    for code in ("KR", "US"):
        raw = _mapping(markets.get(code)) or _mapping(scanned.get(code))
        market_rows[code] = {
            "market": code,
            "phase": raw.get("phase") or raw.get("session_phase") or "UNKNOWN",
            "allows_new_entry": raw.get("allows_new_entry"),
            "reason": raw.get("reason") or raw.get("detail"),
            "regime": raw.get("regime"),
            "source": raw.get("source"),
            "sources": raw.get("sources") or [],
            "quote_age_seconds": raw.get("quote_age_seconds"),
            "healthy_symbols": raw.get("healthy_symbols"),
            "strategy_ids": raw.get("strategy_ids") or raw.get("strategies") or [],
            "updated_at": raw.get("updated_at"),
        }
    pipeline = []
    blocked = False
    for link in chain:
        # The diagnostic checks may run independently: do not label a later
        # successful result 'unevaluated' solely because an earlier gate failed.
        state = "passed" if link.get("ok") is True else "blocked"
        pipeline.append({
            "stage": link.get("stage"),
            "label": STAGE_LABELS.get(link.get("stage"), link.get("stage") or "진단"),
            "state": state,
            "first_blocker": state == "blocked" and not blocked,
            "detail": link.get("detail") or "",
        })
        blocked = blocked or state == "blocked"
    armed_link = next((item for item in chain if item.get("stage") == "live_armed"), {})
    blocker = {
        "stage": first_blocker.get("stage") if first_blocker else blockade.get("blocking_stage"),
        "detail": first_blocker.get("detail") if first_blocker else blockade.get("blocking_detail"),
    }
    runtime = _mapping(runtime)
    return {
        "generated_at": (now or datetime.now(timezone.utc)).isoformat(),
        "markets": market_rows,
        "trading": {
            "running": trading.get("running", status.get("running")),
            "mode": mode if isinstance(mode, str) else _mapping(_mapping(mode).get("active")).get("mode"),
            "armed": armed_link.get("ok"),
            "trading_possible": bool(chain)
            and trading.get("running", status.get("running")) is True
            and armed_link.get("ok") is True
            and blockade.get("ok") is not False
            and all(item.get("ok") is True for item in chain),
            "last_cycle_at": status.get("last_cycle_at") or session.get("last_cycle_at"),
            "selected_symbol": session.get("selected_symbol"),
            "selected_strategy": session.get("selected_strategy"),
            "phase": session.get("phase"),
            "execution_scope": "single_account_owner",
            "blocker": blocker,
            "pipeline": pipeline,
            "orders": _order_rows(orders or []),
        },
        "runtime": {key: runtime.get(key) for key in (
            "backend", "requested_backend", "device", "latency_ms", "model",
            "fallback_reason", "learning", "host", "state_path",
        )},
    }


OPERATIONS_PAGE = """<!doctype html>
<html lang="ko">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="color-scheme" content="dark">
  <title>OBAITS · 트레이딩 현황</title>
  <link rel="icon" href="/static/icon.png">
  <link rel="stylesheet" href="/static/operations_dashboard.css?v=20260920">
  <script defer src="/static/operations_dashboard.js?v=20260920"></script>
</head>
<body>
  <a class="skip-link" href="#profit">손익으로 이동</a>
  <main class="dashboard">
    <header class="page-header">
      <div class="brand"><span class="brand-icon" aria-hidden="true">O</span><div><p class="eyebrow">OBAITS / TRADING DESK</p><h1>트레이딩 현황</h1></div></div>
      <nav aria-label="운영 메뉴"><a href="/account/advanced">전략 상세</a><a href="/display/ontology">관계 그래프</a><button id="refresh" type="button">새로고침</button></nav>
    </header>
    <div class="connection-bar"><span id="connection" role="status">상태 확인 중</span><span id="updated">—</span></div>

    <section id="profit" aria-labelledby="profit-title">
      <div class="section-heading"><h2 id="profit-title">계좌와 손익</h2><span id="account-freshness" class="badge">확인 대기</span></div>
      <div class="kpis">
        <article class="kpi primary"><h3>오늘 실현 손익</h3><strong id="pnl-today">—</strong><p>계좌 기준 · 체결된 거래</p></article>
        <article class="kpi"><h3>평가 손익</h3><strong id="pnl-unrealized">—</strong><p>현재 보유 종목</p></article>
        <article class="kpi"><h3>총자산</h3><strong id="assets">—</strong><p id="assets-detail">국내·해외 합산 · 원화 환산</p></article>
        <article class="kpi"><h3>주문 가능 현금</h3><strong id="cash-krw">—</strong><p id="cash-usd">USD —</p></article>
      </div>
      <p class="source-note" id="account-note">마지막으로 확인된 실계좌 데이터를 불러옵니다.</p>
    </section>

    <section aria-labelledby="markets-title">
      <div class="section-heading"><h2 id="markets-title">지금의 시장</h2><p>국내·미국 독립 관측</p></div>
      <div class="markets" id="markets"><article class="card empty">시장 상태 확인 중</article></div>
    </section>

    <section class="card execution" aria-labelledby="execution-title">
      <div class="section-heading"><h2 id="execution-title">거래는 어디까지 진행됐나요?</h2><span class="badge" id="execution-mode">확인 대기</span></div>
      <div class="verdict" id="verdict" role="status"><span class="verdict-indicator" aria-hidden="true"></span><div><h3 id="blocker-title">주문 경로 확인 중</h3><p id="blocker-detail">엔진 상태와 진입 조건을 확인합니다.</p></div></div>
      <ol class="pipeline" id="pipeline"></ol>
      <div class="execution-footer"><span id="execution-owner">선택된 전략 확인 대기</span><span>시장별 관측 · 계좌당 한 실행 주체</span></div>
    </section>

    <div class="lower-grid">
      <section class="card" aria-labelledby="holdings-title">
        <div class="section-heading"><h2 id="holdings-title">보유 종목</h2><span id="holding-count">—</span></div>
        <div class="table-scroll"><table><thead><tr><th scope="col">종목 / 시장</th><th scope="col">수량</th><th scope="col">현재가</th><th scope="col">평가 손익</th></tr></thead><tbody id="holdings"><tr><td colspan="4" class="empty">계좌 확인 대기</td></tr></tbody></table></div>
      </section>
      <section class="card" aria-labelledby="runtime-title">
        <div class="section-heading"><h2 id="runtime-title">추론과 모델</h2><span class="badge" id="backend-badge">확인 대기</span></div>
        <dl class="runtime-list" id="runtime"><div><dt>실제 추론 장치</dt><dd>확인 대기</dd></div></dl>
        <p class="source-note" id="runtime-note">실행 중인 장치와 검증된 모델을 표시합니다.</p>
      </section>
    </div>

    <section class="card" aria-labelledby="orders-title">
      <div class="section-heading"><h2 id="orders-title">최근 주문과 체결</h2><span id="order-count">—</span></div>
      <div class="table-scroll"><table><thead><tr><th scope="col">시각</th><th scope="col">종목</th><th scope="col">방향</th><th scope="col">주문 / 체결 수량</th><th scope="col">상태</th></tr></thead><tbody id="orders"><tr><td colspan="5" class="empty">주문 이력 확인 대기</td></tr></tbody></table></div>
    </section>
    <footer class="page-footer"><span>현금 주식 중심 · 관측과 체결을 구분합니다.</span><a href="/account/advanced">차트·학습·상세 진단 열기 ↗</a></footer>
  </main>
</body>
</html>"""

File: src/app/static/operations_dashboard.js
/* Read-only cached monitoring. Missing observations must never look like zero P&L
   or live order permission. No broker refresh, POST, WebGL, or animation loop. */
(() => {
  'use strict';
  const state = { busy: false, timer: null, operations: null, account: null, failures: [], lastSuccess: null };
  const $ = (id) => document.getElementById(id);
  const text = (id, value) => { const element = $(id); if (element) element.textContent = value ?? '—'; };
  const escape = (value) => String(value ?? '').replace(/[&<>"']/g, (char) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;' }[char]));
  const numeric = (value) => value === null || value === undefined || value === '' ? null : Number.isFinite(Number(value)) ? Number(value) : null;
  const number = (value, digits = 0) => numeric(value) === null ? '—' : numeric(value).toLocaleString('ko-KR', { maximumFractionDigits: digits });
  const money = (value) => numeric(value) === null ? '—' : `${number(value)}원`;
  const clock = (value) => !value || Number.isNaN(new Date(value).getTime()) ? '확인 대기' : new Date(value).toLocaleString('ko-KR', { month: '2-digit', day: '2-digit', hour: '2-digit', minute: '2-digit', second: '2-digit', hour12: false });
  const signed = (value) => numeric(value) > 0 ? `+${money(value)}` : money(value);
  const tone = (value) => numeric(value) > 0 ? 'positive' : numeric(value) < 0 ? 'negative' : '';
  const phases = { UNKNOWN: '확인 대기', REGULAR: '정규장', PRE: '프리마켓', PREMARKET: '프리마켓', AFTER: '애프터마켓', AFTERHOURS: '애프터마켓', CLOSED: '휴장', DAY: '주간 거래', DAYTIME: '주간 거래', AUCTION: '동시호가', NXT: 'NXT', KR_REGULAR: '정규장', US_REGULAR: '정규장', KR_CLOSED: '휴장', US_CLOSED: '휴장', US_PREMARKET: '프리마켓', US_AFTERHOURS: '애프터마켓', US_DAYTIME: '주간 거래' };
  Object.assign(phases, { KRX_PREOPEN: 'KRX 장전', KRX_OPENING_AUCTION: 'KRX 시초 동시호가', KRX_REGULAR: 'KRX 정규장', KRX_CLOSING_AUCTION: 'KRX 종가 동시호가', KRX_AFTER_CLOSE: 'KRX 장후', KRX_AFTER_SINGLE_PRICE: 'KRX 시간외 단일가', NXT_PRE: 'NXT 프리마켓', NXT_REGULAR: 'NXT 정규장', NXT_POST: 'NXT 애프터마켓', US_AFTERMARKET: '애프터마켓' });
  const regimes = { bull: '상승', bullish: '상승', trend_up: '상승 추세', bear: '하락', bearish: '하락', trend_down: '하락 추세', sideways: '횡보', range: '횡보', volatile: '변동성 확대', high_volatility: '변동성 확대', high_vol: '변동성 확대', stress: '위험 회피', risk_off: '위험 회피', neutral: '중립', unknown: '확인 대기' };
  const reasons = { MARKET_POLICY_CASH_REGIME: '시장 프로필: 현금 대기 권고', SESSION_LIVE_AUTHORIZATION_REQUIRED: '세션 실주문 권한 확인 필요', SESSION_CLOSED: '현재 거래 가능 세션 없음', NO_ACTIVE_SESSION: '현재 거래 가능 세션 없음' };
  const stages = { engine: '엔진', live_armed: '주문 권한', market_session: '거래 시간', buy_candidates: '후보 종목', micro_buy_intents: '진입 조건', strategy_election: '전략 선택', position: '포지션' };
  const orderStates = { FILLED: '체결', PARTIALLY_FILLED: '일부 체결', ACCEPTED: '접수', SUBMITTED: '제출', REJECTED: '거절', BLOCKED: '차단', CANCELED: '취소', CANCELLED: '취소', OPEN: '대기', WORKING: '대기', NO_ORDER: '주문 없음' };

  function badge(id, value, className = '') { const node = $(id); node.textContent = value; node.className = `badge ${className}`; }
  function renderAccount(account) {
    const snapshot = account?.snapshot || null;
    const checked = snapshot && account?.status !== 'unavailable';
    const stamp = account?.last_verified_at || snapshot?.updated_at;
    const stampMs = Date.parse(stamp || '');
    const fresh = checked && account?.authoritative === true && account?.is_stale !== true && Number.isFinite(stampMs) && Date.now() - stampMs <= 120000 && stampMs <= Date.now() + 5000;
    $('profit').classList.toggle('account-stale', !fresh);
    badge('account-freshness', fresh ? '실계좌 확인값' : checked ? '마지막 확인값' : '계좌 확인 대기', fresh ? 'good' : 'warn');
    for (const [id, key] of [['pnl-today', 'realized_pnl_today_krw'], ['pnl-unrealized', 'unrealized_pnl_krw']]) {
      const value = checked ? snapshot[key] : null;
      text(id, signed(value)); $(id).className = tone(value);
    }
    text('assets', money(checked ? snapshot.total_asset_krw : null));
    const cash = checked ? snapshot.orderable_cash_by_currency || {} : {};
    text('cash-krw', money(cash.KRW)); text('cash-usd', `USD ${number(cash.USD, 2)}`);
    text('account-note', `${fresh ? '실계좌' : checked ? '보관된 실계좌' : '계좌 미확인'} · ${clock(stamp)}${checked && !fresh ? ' · 현재 잔고와 차이가 있을 수 있습니다.' : ''}`);
    const rows = checked && Array.isArray(account.holdings) ? account.holdings : [];
    text('holding-count', checked ? `${rows.length}종목` : '확인 대기');
    $('holdings').innerHTML = rows.length ? rows.map((row) => `<tr><td><strong>${escape(row.name || row.ticker)}</strong><small>${escape(row.ticker)} · ${escape(row.market || row.market_group || '—')}</small></td><td>${number(row.quantity, 4)}</td><td>${number(row.current_price ?? row.last_price, row.currency === 'USD' ? 2 : 0)} ${escape(row.currency || '')}</td><td class="${tone(row.unrealized_pnl_krw)}">${signed(row.unrealized_pnl_krw)}<small>${numeric(row.unrealized_pnl_rate) === null ? '—' : `${number(row.unrealized_pnl_rate * 100, 2)}%`}</small></td></tr>`).join('') : `<tr><td colspan="4" class="empty">${checked ? '보유 중인 종목이 없습니다.' : '확인된 실계좌 정보가 없습니다.'}</td></tr>`;
  }

  function renderMarkets(markets) {
    $('markets').innerHTML = ['KR', 'US'].map((code) => {
      const row = markets?.[code] || {};
      const phase = String(row.phase || 'UNKNOWN').toUpperCase();
      const allowed = row.allows_new_entry;
      const strategies = Array.isArray(row.strategy_ids) ? row.strategy_ids.map((item) => typeof item === 'object' ? item.name || item.strategy_id || '' : item).filter(Boolean) : [];
      const source = row.source || (Array.isArray(row.sources) ? row.sources.join(' · ') : '') || '데이터 출처 확인 대기';
      const summary = row.reason ? String(row.reason).split(/,\s*/).map((item) => reasons[item] || item).join(' · ') : (allowed === true ? '세션 진입 가능 · 종목별 비용과 리스크 조건 평가' : allowed === false ? '현재 세션에서 신규 진입 대기' : '시장 시간과 주문 가능 여부를 확인하고 있습니다.');
      const phaseLabel = phase.split(/,\s*/).map((item) => phases[item] || item).join(' · ');
      return `<article class="card"><div class="market-heading"><div><p class="market-code">${code} / CASH EQUITY</p><h3>${code === 'KR' ? '국내 시장' : '미국 시장'}</h3></div><span class="badge ${allowed === true ? 'good' : allowed === false ? 'warn' : ''}">${escape(phaseLabel)}</span></div><p class="market-summary">${escape(summary)}</p><dl class="market-metrics"><div><dt>시장 국면</dt><dd>${escape(regimes[String(row.regime || 'unknown').toLowerCase()] || row.regime)}</dd></div><div><dt>시세 경과</dt><dd>${numeric(row.quote_age_seconds) === null ? '확인 대기' : `${number(row.quote_age_seconds, 1)}초`}</dd></div><div><dt>신선한 종목</dt><dd>${numeric(row.healthy_symbols) === null ? '확인 대기' : `${number(row.healthy_symbols)}개`}</dd></div></dl><p class="market-source">${escape(source)}${strategies.length ? `<br>전략 · ${escape(strategies.join(' / '))}` : ''}</p></article>`;
    }).join('');
  }

  function renderTrading(trading) {
    const row = trading || {};
    const pipeline = Array.isArray(row.pipeline) ? row.pipeline : [];
    const first = pipeline.find((item) => item.first_blocker) || pipeline.find((item) => item.state === 'blocked');
    const ready = row.trading_possible === true && row.running === true && row.armed === true;
    const blocker = row.blocker || {};
    let title = ready ? '전략 평가 가능' : blocker.stage ? `${stages[blocker.stage] || blocker.stage} · 확인 필요` : row.running === false ? '거래 엔진 대기' : '진입 상태 확인 대기';
    text('blocker-title', title);
    text('blocker-detail', blocker.detail || first?.detail || (ready ? '관측된 평가 조건을 충족했습니다. 종목별 주문 직전 검증이 남아 있으며, 실제 제출과 체결은 아래 이력에서 확인하세요.' : '실행 상태를 확인할 수 있는 진단 정보가 아직 없습니다.'));
    $('verdict').className = `verdict ${ready ? 'good' : ''}`;
    const modes = { live_trading: '실거래', live: '실거래', paper: '모의거래', shadow: '관측', analysis_only: '분석' };
    badge('execution-mode', `${modes[row.mode] || row.mode || '모드 확인 대기'} · ${row.armed === true ? '주문 권한 확인' : row.armed === false ? '주문 잠금' : '권한 확인 대기'}`, row.armed === true ? 'good' : 'warn');
    $('pipeline').innerHTML = pipeline.map((step, index) => `<li class="${step.state === 'passed' ? 'passed' : 'blocked'} ${step.first_blocker ? 'first-blocker' : ''}" title="${escape(step.detail)}"><span class="stage-index">${String(index + 1).padStart(2, '0')}</span><strong>${escape(step.label || stages[step.stage] || step.stage)}</strong><small>${step.state === 'passed' ? '통과' : step.first_blocker ? '첫 대기 지점' : '조건 미충족'}</small></li>`).join('');
    text('execution-owner', row.selected_strategy ? `${row.selected_symbol || '—'} · ${row.selected_strategy} · ${row.phase || '선택됨'}` : '선택된 실행 전략 없음');
    const orders = Array.isArray(row.orders) ? row.orders : [];
    text('order-count', `${orders.length}건`);
    $('orders').innerHTML = orders.length ? orders.slice().reverse().map((order) => `<tr><td>${escape(clock(order.occurred_at || order.recorded_at || order.updated_at))}</td><td>${escape(order.ticker || order.symbol || '—')}</td><td>${escape(({ BUY: '매수', SELL: '매도' })[String(order.side).toUpperCase()] || order.side || '—')}</td><td>${number(order.ordered_quantity ?? order.quantity, 4)} / ${number(order.filled_quantity, 4)}</td><td>${escape(orderStates[order.order_status || order.status] || order.order_status || order.status || '확인 대기')}</td></tr>`).join('') : '<tr><td colspan="5" class="empty">확인된 최근 주문이 없습니다. 신호 발생은 주문 체결과 다릅니다.</td></tr>';
  }

  function renderRuntime(runtime) {
    const row = runtime || {};
    const model = row.model || {};
    const learning = row.learning || {};
    badge('backend-badge', row.backend || '확인 대기', row.backend ? '' : 'warn');
    const predictionStatus = ({ READY: '진입 예측 조건 통과', WAITING: '진입 예측 조건 대기' })[model.status] || model.status;
    const pairs = [['실제 추론 장치', row.device || row.backend], ['요청한 장치', row.requested_backend], ['추론 지연', numeric(row.latency_ms) === null ? null : `${number(row.latency_ms, 2)} ms`], ['모델 버전', model.version], ['모델 갱신', model.updated_at ? clock(model.updated_at) : null], ['최근 예측 판정', predictionStatus], ['학습 상태', learning.status || (learning.active === true ? '학습 중' : learning.active === false ? '대기' : null)]];
    $('runtime').innerHTML = pairs.map(([label, value]) => `<div><dt>${escape(label)}</dt><dd>${escape(value || '확인 대기')}</dd></div>`).join('');
    text('runtime-note', row.fallback_reason || model.reason || '관측된 실행 장치 기준입니다. 새 모델은 검증 결과에 따라 반영됩니다.');
  }

  async function fetchJson(url) {
    const controller = new AbortController();
    const timeout = window.setTimeout(() => controller.abort(), 8000);
    try {
      const response = await fetch(url, { cache: 'no-store', credentials: 'same-origin', signal: controller.signal });
      if (!response.ok) throw new Error(`HTTP ${response.status}`);
      const payload = await response.json();
      if (!payload || typeof payload !== 'object' || Array.isArray(payload)) throw new Error('Invalid dashboard response');
      return payload;
    } finally { window.clearTimeout(timeout); }
  }

  async function refresh() {
    if (state.busy || document.hidden) return;
    state.busy = true; $('refresh').disabled = true;
    try {
      const results = await Promise.allSettled([fetchJson('/api/operations/overview'), fetchJson('/api/account/summary')]);
      state.failures = [];
      results.forEach((result, index) => {
        if (result.status === 'fulfilled') state[index === 0 ? 'operations' : 'account'] = result.value;
        else state.failures.push(index === 0 ? '운영 상태' : '계좌');
      });
      if (results[0].status === 'fulfilled') {
        renderMarkets(state.operations.markets); renderTrading(state.operations.trading); renderRuntime(state.operations.runtime);
        state.lastSuccess = state.operations.generated_at || new Date().toISOString();
      }
      renderAccount(state.failures.includes('계좌') && state.account ? { ...state.account, authoritative: false, is_stale: true } : state.account);
      text('updated', `운영 상태 갱신 ${clock(state.lastSuccess)}`);
      text('connection', state.failures.length ? `${state.failures.join(' · ')} 연결 지연 · 마지막 확인값 표시` : '연결됨 · 5초 간격 확인');
      $('connection').className = state.failures.length ? 'stale' : '';
      if (state.failures.includes('운영 상태')) {
        $('verdict').className = 'verdict';
        text('blocker-title', '운영 상태 연결 지연');
        text('blocker-detail', '마지막 진단을 표시하고 있습니다. 현재 주문 권한과 시장 상태는 확인할 수 없습니다.');
        badge('execution-mode', '연결 지연 · 현재 권한 확인 불가', 'warn');
      }
    } finally { state.busy = false; $('refresh').disabled = false; }
  }
  function schedule() { window.clearTimeout(state.timer); state.timer = window.setTimeout(async () => { await refresh(); schedule(); }, 5000); }
  // Small pure formatting hooks support offline contract tests without a server.
  if (typeof module !== 'undefined' && module.exports) module.exports = { numeric, number, money, escape, renderAccount, renderMarkets, renderTrading, renderRuntime, refresh, state };
  if (typeof document !== 'undefined' && $('refresh')) {
    $('refresh').addEventListener('click', refresh);
    document.addEventListener('visibilitychange', () => { if (!document.hidden) refresh(); });
    refresh(); schedule();
  }
})();
